_posts/2016-10-08-eidolon-deploy-process.markdown

```diff
@@ -0,0 +1,192 @@
+---
+layout: post
+title: "The Eidolon Deploy Process"
+date: 2016-10-08 12:00
+author: ash
+categories: [mobile, eidolon]
+---
+
+Since we [originally built](http://artsy.github.io/blog/2014/11/13/eidolon-retrospective/) Eidolon – an auction bidding kiosk app – the project has largely remained in maintenance mode. Eidolon was one of the first projects that we used automated deploys for, and the deploy process has remained largely unchanged. I believe this stability of the deploy process is a testament to how well the automated deploys have gone. 
+
+This post is going to detail the mechanics of automated deploys for an enterprise-distributed iOS application, discuss lessons we learned and applied to other projects' deploy processes, and describe some of the changes we'd like to make. Our project is entirely open source, so you can check out any part of the code on your own or [open an issue](https://github.com/artsy/eidolon/issues/new) with questions.
+
+<!-- more -->
+
+## Deploying Eidolon
+
+It's one command on the terminal to deploy Eidolon:
+
+```sh
+bundle exec fast lane deploy version:X.Y.Z
+```
+
+This command does a lot of things. It uses [Fastlane](https://fastlane.tools), and you can [read the entire script here](https://github.com/artsy/eidolon/blob/a0aad31bccfe2b4abf648fc64892cc165be400b4/fastlane/Fastfile#L40-L131). We're going to go over each part line-by-line. A few notes:
+
+- We run this command locally on a development machine that has the keys installed to sign a deploy.
+- Our changelog is formatted in [yaml](https://en.wikipedia.org/wiki/YAML), our script uses this strategically.
+- Our deploy script modifies the project's Info.plist version and build number, as well as the changelog.
+
+Let's dive in!
+
+## The Script
+
+The first thing we do is verify that the version number we've been given is in the proper [semver](http://semver.org) format.
+
+```rb
+version = options[:version]
+raise "You must specify a version in A.B.X format to deploy." if version.nil? || version.scan(/\d+\.\d+\.\d+/).length == 0
+```
+
+We deploy using Hockey, so make sure that an environment variable with the Hockey API key is set.
+
+```rb
+hockey_api_token = ENV['HOCKEY_API_TOKEN']
+raise "You must specify a HOCKEY_API_TOKEN environment variable to deploy." if hockey_api_token.nil?
+```
+
+We also want to verify that we have valid Artsy API keys
+
+```rb
+  verify_pod_keys
+  ensure_git_status_clean
+```
+
+Next we need to set the build number. These need to be unique, and we use the current date. This could be a problem if we need to deploy more than once in a day. It hasn't been a problem yet, though, since we rarely deploy. 
```

Comment:
if you care, this has been a problem with Eigen, and so we deploy with the hour too